pkg/virt-handler/node-labeller/node_labeller.go

```diff
@@ -231,31 +228,63 @@ func skipNodeLabelling(node *v1.Node) bool {
 	return exists
 }
 
+func replaceLabels(labels map[string]string) patch.PatchOperation {
+	return patch.PatchOperation{
+		Op:    "replace",
+		Path:  "/metadata/labels",
+		Value: labels,
+	}
+}
+
+func replaceAnnotations(annotations map[string]string) patch.PatchOperation {
+	return patch.PatchOperation{
+		Op:    "replace",
+		Path:  "/metadata/annotations",
+		Value: annotations,
+	}
+}
+
 func (n *NodeLabeller) patchNode(originalNode, node *v1.Node) error {
 	p := make([]patch.PatchOperation, 0)
+	shadownodeOperation := make([]patch.PatchOperation, 0)
 	if !equality.Semantic.DeepEqual(originalNode.Labels, node.Labels) {
+		// TODO: This can be safely removed once we do not support upgrade from a version that does not have Shadow Node
+		now, err := json.Marshal(metav1.Now())
+		if err != nil {
+			log.DefaultLogger().Reason(err).Errorf("Can't determine date")
+			return err
+		}
 		p = append(p, patch.PatchOperation{
 			Op:    "test",
 			Path:  "/metadata/labels",
 			Value: originalNode.Labels,
-		}, patch.PatchOperation{
-			Op:    "replace",
-			Path:  "/metadata/labels",
-			Value: node.Labels,
-		})
+		}, replaceLabels(node.Labels))
+
+		if len(node.Annotations) == 0 {
+			node.Annotations = map[string]string{}
+		}
+		node.Annotations["kubevirt.io/shadownode"] = string(now)
+		shadownodeOperation = append(shadownodeOperation, replaceLabels(node.Labels))
```

Comment:
Why here you don't need originalNode.Labels like in line 237 for node? Is that because you assume you can do a replace here without check?

Reply:
Good question, I think this is wrong as shadownode could have more information than node and I could replace it/lose it.
I think I need to rework this in a way that I individually construct `add`/`replace`/`delete` patches.